### PR TITLE
MNIST dataset reader implementation

### PR DESCRIPTION
**Pull request by reyoung**

### CODE REVIEW

File: python/paddle/v2/data_set/mnist.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
+import sklearn.datasets.mldata
+import sklearn.model_selection
+import numpy
+
+__all__ = ['MNISTReader', 'train_reader_creator', 'test_reader_creator']
+
+DATA_HOME = None
```
Comment (reyoung) [inline python/paddle/v2/data_set/mnist.py]: @wangkuiyi 修改这个变量可以指定数据集下载的目录。

不过要如issue #1419 中描述，要将下载目录改成`/usr/local`似乎略有不妥。因为`/usr/local`目录一般需要root权限才能读写。常见不需要root权限的目录就是 $HOME 目录。 `/tmp`目录也不太合适。因为重启就没了，并且有的机器`/tmp`目录会映射到 shared_memory 中，并不会很大。

Comment (wangkuiyi) [inline python/paddle/v2/data_set/mnist.py]: 是的。确实应该在 $HOME 里比较方便。

Comment (wangkuiyi) [inline python/paddle/v2/data_set/mnist.py]: 另外，这个变量是所有module都会用到的。是不是需要放到 __init__.py 或者什么其他地方，而不是在mnist.py里呢？

File: python/paddle/v2/data_set/mnist.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    mnist dataset reader. The `train_reader` and `test_reader` method returns
+    a iterator of each sample. Each sample is combined by 784-dim float and a
+    one-dim label
+    """
+
+    def __init__(self, random_state):
+        data = sklearn.datasets.mldata.fetch_mldata(
+            "MNIST original", data_home=DATA_HOME)
+        n_train = 60000
+        self.X_train, self.X_test, self.y_train, self.y_test = sklearn.model_selection.train_test_split(
```
Comment (wangkuiyi) [inline python/paddle/v2/data_set/mnist.py]: 这个数据集有多大？这样全部load进内存合适吗？

File: python/paddle/v2/data_set/mnist.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    """
+    mnist dataset reader. The `train_reader` and `test_reader` method returns
+    a iterator of each sample. Each sample is combined by 784-dim float and a
+    one-dim label
+    """
+
+    def __init__(self, random_state):
+        data = sklearn.datasets.mldata.fetch_mldata(
+            "MNIST original", data_home=DATA_HOME)
+        n_train = 60000
```
Comment (wangkuiyi) [inline python/paddle/v2/data_set/mnist.py]: 这里的60000可能改成 total data size 以及 train/test split ratio 更合适吧？

File: python/paddle/v2/data_set/mnist.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+            random_state=random_state)
+
+    def train_reader(self):
+        return __mnist_reader__(self.X_train, self.y_train)
+
+    def test_reader(self):
+        return __mnist_reader__(self.X_test, self.y_test)
+
+
+__default_instance__ = MNISTReader(0)
```
Comment (wangkuiyi) [inline python/paddle/v2/data_set/mnist.py]: 看上去这里不需要一个class MNISTReader。而是可以把 MNISTReader.__init__ 定义成一个全局函数。这样也不需要 trainer_reader_creator 调用 MNISTReader.train_reader ，以及 test_reader_creator 调用 MNISTReader.test_reader了。

**Comment by wangkuiyi** [comment]
I rewrote the program without using class. Two version -- with/without classes -- are as below:

screen shot 2017-02-23 at 12 06 42 pm

It seems that to realize exactly the same functionality, the version without class is shorter. 

Here follows the shorter version in text format:

```python
import sklearn.datasets.mldata
import sklearn.model_selection
import numpy

__all__ = ['train', 'test']

DATA_HOME = None
DATA_INSTANCES = 60000

def load_and_split_data(random_state):
    data = sklearn.datasets.mldata.fetch_mldata(
        "MNIST original", data_home=DATA_HOME)
    return sklearn.model_selection.train_test_split(
        data.data / 255.0,
        data.target.astype("int"),
        train_size=DATA_INSTANCES,
        random_state=random_state)

X_train, X_test, y_train, y_test = load_and_split_data(0)

def reader_creator(data, target):
    def reader():
        n_samples = data.shape[0]
        for i in xrange(n_samples):
            yield data[i].astype(numpy.float32), int(target[i])
    return reader

def train():
    """
    Default train set reader creator.
    """
    return reader_creator(X_train, y_train)


def test():
    """
    Default test set reader creator.
    """
    return reader_creator(X_test, y_test)


def unittest():
    assert len(list(train()())) == DATA_INSTANCES


if __name__ == '__main__':
    unittest()
```

File: python/paddle/v2/data_set/mnist.py (hunk 1 above)
Comment (reyoung) [inline python/paddle/v2/data_set/mnist.py]: > 另外，这个变量是所有module都会用到的。是不是需要放到 init.py 或者什么其他地方，而不是在mnist.py里呢？

确实，这个变量应该放到一个global的地方，让这个包里的其他实现都可以用上。我这就改一下。

File: python/paddle/v2/data_set/mnist.py (hunk 3 above)
Comment (reyoung) [inline python/paddle/v2/data_set/mnist.py]: > 这里的60000可能改成 total data size 以及 train/test split ratio 更合适吧？

OK, 有道理。

File: python/paddle/v2/data_set/mnist.py (hunk 2 above)
Comment (reyoung) [inline python/paddle/v2/data_set/mnist.py]: > 这个数据集有多大？这样全部load进内存合适吗？

这个数据集非常小。大小为 70000 张 28*28的灰度图。内存占用约为60MB左右。

File: python/paddle/v2/data_set/mnist.py (hunk 4 above)
Comment (reyoung) [inline python/paddle/v2/data_set/mnist.py]: `MNISTReader`可以预留出来，让用户自定义mnist数据集shuffle，split工作(这应该是必要的)。这对于用户测试模型，特别是类似于实现 `cross-validation`的情况下，能够稍微自定义split的方式比较有**必要**。

并且，对比了二者实现，从行数上来说，使用`class`的实现也可以减少。两种实现策略，除去注释和空行的实现行数差别，应该不超过三四行。

例如，class实现中`train_reader_creator`那几行纯粹是为了写注释，否则直接一行

```python
train_reader_creator=__default_instance__.train_reader
```

就搞定了。

Comment (wangkuiyi) [inline python/paddle/v2/data_set/mnist.py]: shuffle应该交由用户使用 @helinwang 写的 paddle.reader.shuffle 来做。不能在用户不知情的情况下替他shuffle，否则每两次训练，结果都不一样。

这里我重写了你的代码，指出两种方式的差别，不仅代码行数不同，而且需要用户了解的背景知识也可以减少，但是你就这么两句话给驳回了？ 不必要的代码，即便只多出去一行，我们也应该删掉。

File: python/paddle/v2/data_set/mnist.py (hunk 2 above)
Comment (wangkuiyi) [inline python/paddle/v2/data_set/mnist.py]: OK

**Comment by wangkuiyi** [comment]
- https://github.com/PaddlePaddle/Paddle/pull/1438#discussion_r102799972
- https://github.com/PaddlePaddle/Paddle/pull/1438#discussion_r103038151

File: python/paddle/v2/data_set/mnist.py (hunk 4 above)
Comment (reyoung) [inline python/paddle/v2/data_set/mnist.py]: Done.

**Comment by reyoung** [comment]
Merge it firstly, because it block the `test` development.
